leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java

```diff
@@ -50,7 +51,7 @@
 
 public class PaxosStateLogImpl<V extends Persistable & Versionable> implements PaxosStateLog<V> {
 
-    private final ReentrantLock lock = new ReentrantLock();
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
     private final Map<Long, Long> seqToVersionMap = Maps.newHashMap();
```

Comment:
Hmm. Could we put `@GuardedBy("lock.writeLock()")`, or more defensively and probably better make this a concurrent hash map? Right now it's fine, but if later on a reader tries to use this (e.g. for caching) then memory visibility becomes a bit tricky.